src/ORM/DataObject.php

```diff
@@ -318,88 +337,107 @@ public static function getSchema()
     /**
      * Construct a new DataObject.
      *
-     * @param array|null $record Used internally for rehydrating an object from database content.
-     *                           Bypasses setters on this class, and hence should not be used
-     *                           for populating data on new records.
-     * @param boolean $isSingleton This this to true if this is a singleton() object, a stub for calling methods.
-     *                             Singletons don't have their defaults set.
+     * @param array $record Initial record content, or rehydrated record content, depending on $creationType
+     * @param int|boolean $creationType Set to DataObject::CREATE_OBJECT, DataObject::CREATE_HYDRATED, or DataObject::CREATE_SINGLETON. Used by SilverStripe internals as best left as the default by regular users.
      * @param array $queryParams List of DataQuery params necessary to lazy load, or load related objects.
      */
-    public function __construct($record = null, $isSingleton = false, $queryParams = [])
+    public function __construct($record = [], $creationType = self::CREATE_OBJECT, $queryParams = [])
     {
         parent::__construct();
 
-        // Set query params on the DataObject to tell the lazy loading mechanism the context the object creation context
-        $this->setSourceQueryParams($queryParams);
-
-        // Set the fields data.
-        if (!$record) {
-            $record = [
-                'ID' => 0,
-                'ClassName' => static::class,
-                'RecordClassName' => static::class
-            ];
+        // Legacy $record default
+        if ($record === null) {
+            $record = [];
         }
 
-        if ($record instanceof stdClass) {
-            $record = (array)$record;
+        // Legacy $isSingleton boolean
+        if (is_bool($creationType)) {
```

Comment:
Maybe I'm overthinking this, but if some legacy code calls this method with a non-boolean truthy/falsy value weird things will happen. I'd be tempted to have some strong type guards here, to handle cases where  a value different from `false`, `true`, `CREATE_SINGLETON`, `CREATE_OBJECT` or `CREATE_HYDRATED` is passed.

Reply:
It's an edge-case but easily addressed 👍

Reply:
I've opted for
 * Check !is_int rather than is_bool
 * Throw a warning if it's neither int nor bool
 * Throw a fatal error if it's int but not a proper value

If someone has passed $isSingle = 1042 their code will throw a fatal, but that's technically violating the original API and it's a rare that this API is directly used regardless.